opm/simulators/wells/StandardWellV_impl.hpp

```diff
@@ -2265,12 +2273,42 @@ namespace Opm
     void
     StandardWellV<TypeTag>::
     computeWellRatesWithBhp(const Simulator& ebosSimulator,
-                            const EvalWell& bhp,
+                            const std::vector<Scalar>& B_avg,
```

Comment:
I am trying to rewriting this comment. 

I suggest to split this function to be two functions. 

The first part basically applies to the part that when `iterate` to be `true`. It includes resetting the bhp limit at the beginning, solving the well equations and update the explicit quantities, and the recover the the well controls at the end. Please make a proper name for it. 

The second function will be the original `computeWellRatesWithBhp` unchanged. 

With this two functions, you can make a new function called `computeWellPotentialsWithBhpLimit`.

Reply:
The resetting of the BHP limit in the current form is not very proper, since it discard the fact that there might be other controls included there at the same time. It did not cause problem is because it is only used when `iterate` to be `true`. 

And also, we have two places resetting the BHP limits. The other places is the `computeWellPotentials` in `BlackoilWellModel`. We should not need to do it twice. 

I will suggest we only do it in one places.  The well control only contains a BHP limit can be created using some function as follows, 
```
    template<typename TypeTag>
    struct WellControls*
    WellInterface<TypeTag>::
    creatingWellControlWithBhpTargetOnly(const double bhp)
    {
        struct WellControls* controls = well_controls_create();
        const double invalid_alq = -1.e100;
        const double invalid_vfp = -1.;
        well_controls_add_new(BHP, bhp, invalid_alq, invalid_vfp, NULL, controls);

        assert(well_controls_get_num(controls) == 1);
        well_controls_set_current(controls, 0);
        return controls;
    }
```

Then we can do the resetting the well control in some manner as follows,
```
        // back up the old control data
        const struct WellControls* original_well_controls = well_controls_;
        const int original_current_control = well_controls_get_current(original_well_controls);
```

then 
```
well_controls_ = this->creatingWellControlWithBhpTargetOnly(bhp);
```

At the end, we recover the old control

```
        // recovering the old well control
        well_controls_destroy(well_controls_);
        well_controls_ = original_well_controls;
        well_controls_set_current(well_controls_, original_current_control);
```

All these should be the first function mentioned above.  We do not need to do it in the `computeWellPotentials` in `BlackoilWellModel`.

Reply:
In MultisegmentWell, the suggestions will be similar. 

Ideally, we should not need to recreate the `well_container`. I do see that it might needs less cleaning and recovering work by creating a new one.

Reply:
Great you find that the well potentials can be more correct when updating the explicit quantities after solving the system til converged under the desired BHP target. 

It is possible in the future we should always do that, while let us do it step by step.